src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6642,46 +6689,53 @@ private ImmutableArray<FlowAnalysisAnnotations> GetParametersAnnotations(Immutab
         }
 
         private VisitArgumentResult VisitArgumentEvaluate(BoundExpression argument, RefKind refKind, FlowAnalysisAnnotations annotations)
+        {
+            var savedState = VisitArgumentEvaluatePrologue(argument);
+            Visit(argument);
+            return VisitArgumentEvaluateEpilogue(argument, savedState, refKind, annotations);
+        }
+
+        private Optional<LocalState> VisitArgumentEvaluatePrologue(BoundExpression argument)
         {
             Debug.Assert(!IsConditionalState);
-            var savedState = (argument.Kind == BoundKind.Lambda) ? this.State.Clone() : default(Optional<LocalState>);
+            return (argument.Kind == BoundKind.Lambda) ? this.State.Clone() : default(Optional<LocalState>);
```

Comment:
>(argument.Kind == BoundKind.Lambda) [](http://example.com/codeflow?start=19&length=35)

Rather than extracting a helper that clones the state and then checking for the fact of cloning, consider extracting a helper that simply returns a boolean (whether the clon is needed). This also will make it clear that the helper doesn't visit anything, whereas the "VisitArgumentEvaluatePrologue" name sounds like the method might visit something